src/nupic/algorithms/TemporalMemory.cpp

```diff
@@ -0,0 +1,1093 @@
+  set<Cell>& prevWinnerCells,
+  Connections& _connections)
+{
+  set<Cell> _activeCells;
+  set<Cell> _winnerCells;
+  vector<Segment> _learningSegments;
+
+  vector<UInt> _unpredictedColumns(activeColumns.begin(), activeColumns.end());
+
+  if (predictedColumns.size() > 0)
+  {
+    // Resize to the worst case usage
+    _unpredictedColumns.resize(activeColumns.size() + predictedColumns.size());
+
+    // Remove the predicted columns from the 
+    // currently active columns
+    vector<UInt>::iterator it = set_difference(
+      activeColumns.begin(), activeColumns.end(),
+      predictedColumns.begin(), predictedColumns.end(),
+      _unpredictedColumns.begin());
+
+    // Trim remainer of set
+    _unpredictedColumns.resize(it - _unpredictedColumns.begin());
+  }
+
+  for (Int column : _unpredictedColumns)
+  {
+    Segment* bestSegment = NULL;
+    Cell* bestCell = NULL;
+
+    vector<Cell> cells = cellsForColumn(column);
+
+    for (auto cell : cells)
+      _activeCells.insert(cell);
+
+    tie(bestCell, bestSegment) =
+      bestMatchingCell(cells, prevActiveCells, _connections);
+
+    if (bestCell) _winnerCells.insert(*bestCell);
+
+    if (bestSegment == NULL && prevWinnerCells.size())
+    {
+      bestSegment = new Segment(_connections.createSegment(*bestCell));
+    }
+
+    if (bestSegment != NULL)
+    {
+      _learningSegments.push_back(*bestSegment);
+    }
+  }
+
+  return make_tuple(_activeCells, _winnerCells, _learningSegments);
+}
+
+void TemporalMemory::learnOnSegments(
+  vector<Segment>& prevActiveSegments,
+  vector<Segment>& learningSegments,
+  set<Cell>& prevActiveCells,
+  set<Cell>& winnerCells,
+  set<Cell>& prevWinnerCells,
+  Connections& _connections)
+{
+  vector<Segment> _allSegments;
+
+  for (auto segment : prevActiveSegments)
+    _allSegments.push_back(segment);
+  for (auto segment : learningSegments)
+    _allSegments.push_back(segment);
+
+  for (Segment segment : _allSegments)
+  {
+    bool isLearningSegment = (find(
+      learningSegments.begin(), learningSegments.end(),
+      segment) != learningSegments.end());
+
+    bool isFromWinnerCell = (find(
+      winnerCells.begin(), winnerCells.end(),
+      segment.cell) != winnerCells.end());
+
+    vector<Synapse> activeSynapses(activeSynapsesForSegment(
+      segment, prevActiveCells, _connections));
+
+    if (isLearningSegment || isFromWinnerCell)
+    {
+      adaptSegment(segment, activeSynapses, _connections);
+    }
+
+    if (isLearningSegment)
+    {
+      Int n = maxNewSynapseCount_ - Int(activeSynapses.size());
+
+      set<Cell> cells = pickCellsToLearnOn(
+                          n, segment, prevWinnerCells, _connections);
+      for (Cell presynapticCell : cells)
+      {
+        _connections.createSynapse(segment, presynapticCell, initialPermanence_);
+      }
+    }
+  }
+}
+
+tuple<vector<Segment>, set<Cell>> TemporalMemory::computePredictiveCells(
+  set<Cell>& _activeCells,
+  Connections& _connections)
```

Comment:
I have this change locally. Will compare output logs in the morning, and try the other change below. All tests are still passing, but there's obvious divergence in synaptic creation in the tutorial tests, and subsequent extensive test statistics.